File: Python/CreationalDesignPattern/FactoryDesignPattern.py
from abc import ABC, abstractmethod
from typing import Dict, Optional


class INotification(ABC):
    @abstractmethod
    def send(self):
        pass


class SMSNotification(INotification):
    def send(self):
        raise NotImplementedError("Unimplemented method 'send'")


class WhatsAppNotification(INotification):
    def send(self):
        raise NotImplementedError("Unimplemented method 'send'")


# Problem
# class NotificationService:
#     def send_notification(self, type, message):
#         notification = None
#         if type == "SMS":
#             notification = SMSNotification()
#         elif type == "whatsapp":
#             notification = WhatsAppNotification()


class NotificationService:
    def send_notification(self, type: str, message: str):
        notification = NotificationFactory.get_instance(type)
        # validation
        # 10000
        notification.send()


class NotificationFactory:
    registry: Dict[str, INotification] = {}

    @staticmethod
    def get_instance(type: str) -> Optional[INotification]:
        if type == "SMS":
            NotificationFactory.registry.setdefault(type, SMSNotification())
        elif type == "whatsapp":
            NotificationFactory.registry.setdefault(type, WhatsAppNotification())
        elif type == "Slack":
            pass
        return NotificationFactory.registry.get(type)


class FactoryDesignPattern:
    pass

File: C++/CreationalDesignPattern/FactoryDesignPattern.cpp
#include <memory>
#include <stdexcept>
#include <string>
#include <unordered_map>

struct INotification {
    virtual void send() = 0;
    virtual ~INotification() = default;
};

struct SMSNotification : INotification {
    void send() override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct WhatsAppNotification : INotification {
    void send() override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

// Problem
// class NotificationService {
//     void sendNotification(std::string type, std::string message) {
//         std::shared_ptr<INotification> notification = nullptr;
//         if (type == "SMS") {
//             notification = std::make_shared<SMSNotification>();
//         } else if (type == "whatsapp") {
//             notification = std::make_shared<WhatsAppNotification>();
//         }
//     }
// };

class NotificationFactory {
public:
    static std::unordered_map<std::string, std::shared_ptr<INotification>> registry;

    static std::shared_ptr<INotification> getInstance(const std::string& type) {
        if (type == "SMS") {
            registry.emplace(type, std::make_shared<SMSNotification>());
        } else if (type == "whatsapp") {
            registry.emplace(type, std::make_shared<WhatsAppNotification>());
        } else if (type == "Slack") {
        }
        auto it = registry.find(type);
        return it == registry.end() ? nullptr : it->second;
    }
};

std::unordered_map<std::string, std::shared_ptr<INotification>> NotificationFactory::registry;

class NotificationService {
public:
    void sendNotification(const std::string& type, const std::string& message) {
        auto notification = NotificationFactory::getInstance(type);
        // validation
        // 10000
        notification->send();
    }
};

class FactoryDesignPattern {};
